lib/concurrency/Prefix_Refinement.thy

```diff
@@ -1103,36 +1805,273 @@ lemma prefix_refinement_select:
   apply (auto simp add: triv_refinement_def return_def image_def)
   done
 
+lemma prefix_refinement_assert:
+  "P = P' \<Longrightarrow> prefix_refinement sr iosr iosr dc AR R \<top>\<top> \<top>\<top> (assert P) (assert P')"
+  by (simp add: assert_def prefix_refinement_fail prefix_refinement_return_imp)
+
+lemma prefix_refinement_assert_opt:
+  "\<lbrakk>rel_option rvr v v'\<rbrakk>
+   \<Longrightarrow> prefix_refinement sr iosr iosr rvr AR R \<top>\<top> \<top>\<top> (assert_opt v) (assert_opt v')"
+  by (auto simp: assert_opt_def prefix_refinement_fail prefix_refinement_return_imp
+          split: option.splits)
+
+lemma prefix_refinement_assertE:
+  "P = P' \<Longrightarrow> prefix_refinement sr iosr iosr dc AR R \<top>\<top> \<top>\<top> (assertE P) (assertE P')"
+  by (simp add: assertE_def prefix_refinement_fail prefix_refinement_returnOk_imp)
+
+lemma prefix_refinement_gets_the:
+  "\<lbrakk>\<And>s t. \<lbrakk>iosr s t; P s; Q t\<rbrakk> \<Longrightarrow> rel_option rvr (f s) (g t)\<rbrakk>
+   \<Longrightarrow> prefix_refinement sr iosr iosr rvr AR R (\<lambda>_. P) (\<lambda>_. Q) (gets_the f) (gets_the g)"
+  apply (simp add: gets_the_def)
+  apply (rule prefix_refinement_weaken_pre)
+    apply (rule prefix_refinement_bind[where rvr'="rel_option rvr"])
+       apply (rule prefix_refinement_gets[where P=P and Q=Q])
+       apply simp
+      apply (erule prefix_refinement_assert_opt)
+     apply wpsimp+
+  done
+
+\<comment> \<open>FIXME: an equivalent doesn't seem to exist for corres, do we want it to or does it mean we
+          shouldn't bother with this?\<close>
+lemma prefix_refinement_gets_map:
```

Comment:
I think we want it. Projections are even more likely to be useful in the trace monad.